packages/react-components/react-utilities/src/compose/getSlotsNext.ts

```diff
@@ -0,0 +1,47 @@
+import * as React from 'react';
+import type { ComponentState, SlotPropsRecord, UnknownSlotProps } from './types';
+import { ObjectSlotProps, Slots } from './getSlots';
+
+/**
+ * Similar to `getSlots`, main difference is that it's compatible with new custom jsx pragma
+ */
+export function getSlotsNext<R extends SlotPropsRecord>(
```

Comment:
If `getSlotsNext` worked slightly differently (maybe even renamed to `resolveSlotComponents`), it would make the future upgrade to [Option D](https://github.com/microsoft/fluentui/blob/master/rfcs/react-components/convergence/slot-children-render-function.md#option-d-option-a-partially--option-c) (or A) from #27164 be simpler:

* Create another symbol `SLOT_COMPONENT_SYMBOL` to save the value of `state.components.SLOT` on `state.SLOT.SLOT_COMPONENT_SYMBOL` (from Option D).
* `resolveSlotComponents` would only be responsible for setting `SLOT_COMPONENT_SYMBOL` on each slot.
* `jsxFromSlotComponent` would just take in the single slot object (without props being separate: the slot object would have both the component and slots in a single object).

```tsx
const renderFoo = (state: FooState) => {
  resolveSlotComponents(state);

  return (
    <state.root>
      {state.icon && <state.icon />}
      {state.root.children}
    </state.root>
  );
};
```

In the future, we can progressively upgrade components to use the `slot()` function instead of `resolveShorthand()`. The slot() function would set `SLOT_COMPONENT_SYMBOL` at the time the slot is resolved, and the upgraded components would no longer call `resolveSlotComponents()`.